store/sqlstore/thread_store.go

```diff
@@ -52,247 +52,259 @@ func (s *SqlThreadStore) Get(id string) (*model.Thread, error) {
-		}()
+	// If a team is specified, constrain to channels in that team or DMs/GMs without
+	// a team at all.
+	if teamId != "" {
+		query = query.
+			Join("Channels ON Posts.ChannelId = Channels.Id").
+			Where(sq.Or{
+				sq.Eq{"Channels.TeamId": teamId},
+				sq.Eq{"Channels.TeamId": ""},
+			})
+	}
+
+	if !opts.Deleted {
+		query = query.Where(sq.Eq{"Posts.DeleteAt": 0})
 	}
 
-	totalUnreadMentionsResult := <-totalUnreadMentionsChan
-	if totalUnreadMentionsResult.NErr != nil {
-		return nil, totalUnreadMentionsResult.NErr
+	if opts.Since > 0 {
+		query = query.Where(sq.GtOrEq{"ThreadMemberships.LastUpdated": opts.Since})
 	}
-	totalUnreadMentions := totalUnreadMentionsResult.Data.(int64)
 
-	totalCountResult := <-totalCountChan
-	if totalCountResult.NErr != nil {
-		return nil, totalCountResult.NErr
+	if opts.Unread {
+		query = query.Where(sq.Expr("ThreadMemberships.LastViewed < Threads.LastReplyAt"))
 	}
-	totalCount := totalCountResult.Data.(int64)
 
-	totalUnreadThreadsResult := <-totalUnreadThreadsChan
-	if totalUnreadThreadsResult.NErr != nil {
-		return nil, totalUnreadThreadsResult.NErr
+	order := "DESC"
+	if opts.Before != "" {
+		query = query.Where(sq.Expr(`LastReplyAt < (SELECT LastReplyAt FROM Threads WHERE PostId = ?)`, opts.Before))
+	}
+	if opts.After != "" {
+		order = "ASC"
+		query = query.Where(sq.Expr(`LastReplyAt > (SELECT LastReplyAt FROM Threads WHERE PostId = ?)`, opts.After))
 	}
-	totalUnreadThreads := totalUnreadThreadsResult.Data.(int64)
 
-	// userIds is the de-duped list of participant ids from all threads.
-	userIds := []string{}
-	// userIdMap is the map of participant ids from all threads.
-	// Used to generate userIds
-	userIdMap := map[string]bool{}
+	query = query.
+		OrderBy("Threads.LastReplyAt " + order).
+		Limit(pageSize)
 
-	result := &model.Threads{
-		Total:               totalCount,
-		Threads:             []*model.ThreadResponse{},
-		TotalUnreadMentions: totalUnreadMentions,
-		TotalUnreadThreads:  totalUnreadThreads,
+	sql, args, _ := query.ToSql()
+	err := s.GetReplicaX().Select(&threads, sql, args...)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to fetch threads for user id=%s", userId)
 	}
 
-	if !opts.TotalsOnly {
-		threadsResult := <-threadsChan
-		if threadsResult.NErr != nil {
-			return nil, threadsResult.NErr
+	// Build the de-duplicated set of user ids representing participants across all threads.
+	var participantUserIds []string
+	for _, thread := range threads {
+		for _, participantUserId := range thread.Participants {
+			participantUserIds = append(participantUserIds, participantUserId)
 		}
-		threads := threadsResult.Data.([]*JoinedThread)
-		for _, thread := range threads {
-			for _, participantId := range thread.Participants {
-				if _, ok := userIdMap[participantId]; !ok {
-					userIdMap[participantId] = true
-					userIds = append(userIds, participantId)
-				}
-			}
+	}
+	participantUserIds = model.RemoveDuplicateStrings(participantUserIds)
```

Comment:
Just to confirm, we are trading off memory for CPU time?

Reply:
I admit I didn't consider the performance characteristics of this change: I was focussed on simplifying the code, and noticed we already had a helper method to achieve the desired outcome and make the semantics of this block easier to understand. Thoughts?

Reply:
Ah no worries, I was just clarifying. This is fine. :+1: